opencog/atoms/core/ImplicationLink.cc

```diff
@@ -61,26 +61,23 @@ ImplicationLink::ImplicationLink(Link &l)
 void ImplicationLink::extract_variables(const HandleSeq& oset)
 {
 	size_t sz = _outgoing.size();
-	// For now we assume no implicit scoping, so the number of
-	// arguments is 2 when sugar syntax isn't used (no scoping), or 3
-	// when the sugar syntax (scoping) is used.
+
 	if (sz < 2 or 3 < sz)
 		throw InvalidParamException(TRACE_INFO,
-			"Expecting an outgoing set size of at most two, got %d", sz);
+			"Expecting an outgoing set of size 2 or 3, got %d", sz);
+
+	// If there is an explicit variable declaration, then use that.
+	// If there is no variable declaration, then assume that all of
+	// the free variables in the link are scoped.
+	ScopeLink::extract_variables(oset);
```

Comment:
I'd like to remind you that you are the one who added ImplicationLink.cc in the first place ... I was sort of mystified as to why it was needed, or what the point of having it was.  Prior to that addition, there was no special treatment for the ImplicationLink.   Is this class really needed?  Does it really do something important that cannot be done otherwise?

The goal of having C++ classes was not to create or impose semantics, but simply to give a performance boost to something that was going to happen anyway.   Yet ImplicationLink seems to do nothing special for performance, but adds a lot of semantics -- so again, do we even need this C++ class at all? What problem was it meant to solve? Can we just get rid of it (and implement the semantics elsewhere?)